Implement GLFW/OpenAL override by defining and auto-detecting paths to library files

Scrumplex opened the pull request:
Closes PrismLauncher/PrismLauncher#513

Scrumplex committed:
feat: implement override for GLFW/OpenAL with split natives

Fixes PrismLauncher/PrismLauncher#513

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Ryex commented on launcher/minecraft/launch/ExtractNatives.cpp:
```diff
@@ -91,15 +88,16 @@ void ExtractNatives::executeTask()
     auto settings = minecraftInstance->settings();
+
+    // We only need OpenAL here, as modern versions of LWJGL (3+) are handled by JVM args, while older versions (2) didn't have GLFW
     bool nativeOpenAL = settings->get("UseNativeOpenAL").toBool();
-    bool nativeGLFW = settings->get("UseNativeGLFW").toBool();
 
     auto outputPath  = minecraftInstance->getNativePath();
     auto javaVersion = minecraftInstance->getJavaVersion();
     bool jniHackEnabled = javaVersion.major() >= 8;
     for(const auto &source: toExtract)
     {
-        if(!unzipNatives(source, outputPath, jniHackEnabled, nativeOpenAL, nativeGLFW))
+        if(!unzipNatives(source, outputPath, jniHackEnabled, nativeOpenAL))
```
This doesn't break the feature for versions that use lwjgl < 3?

Scrumplex replied:
Did you read the comment a few lines up? :D

Ryex replied:
... the horizontal scroll cut off the end of the comment and I wasn't smart enough to scroll right to see it.

orowith2os commented on launcher/minecraft/MinecraftInstance.cpp:
```diff
@@ -399,6 +399,24 @@ QStringList MinecraftInstance::extraArguments()
         list.append("-Dloader.disable_beacon=true");
     }
+
+#ifdef Q_OS_LINUX
+    {
+        QString openALPath;
+        QString glfwPath;
+
+        if (settings()->get("UseNativeOpenAL").toBool())
+            openALPath = MangoHud::findLibrary("libopenal.so");
+        if (settings()->get("UseNativeGLFW").toBool())
+            glfwPath = MangoHud::findLibrary("libglfw.so");
+
+        if (!openALPath.isEmpty())
+            list.append("-Dorg.lwjgl.openal.libname=" + openALPath);
+        if (!glfwPath.isEmpty())
+            list.append("-Dorg.lwjgl.glfw.libname=" + glfwPath);
+    }
+#endif
```
Would this feature not be potentially useful on other platforms, like Windows and macOS? Or is there some other reason? I know Windows is Not Great in regards to this, is that why?

Scrumplex replied:
Well. There is no implementation for those other platforms yet. This can obviously be expanded in the future, though the main use-case is Linux.

orowith2os commented on launcher/minecraft/launch/ExtractNatives.cpp:
```diff
@@ -91,15 +88,16 @@ void ExtractNatives::executeTask()
     }
     auto settings = minecraftInstance->settings();
+
+    // We only need OpenAL here, as modern versions of LWJGL (3+) are handled by JVM args, while older versions (2) didn't have GLFW
```
```suggestion
    // We only need to manage OpenAL here, as modern versions of LWJGL (3+) are handled by JVM args, and older versions (2) don't have GLFW.
```

Scrumplex committed:
feat: add custom native library path settings

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
fix: remove meta OpenAL workaround

LWJGL 2 doesn't have a separate zip for OpenAL. So there is no reason
for this code.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
fix: fix typo

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
fix: improve debug message

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
I have updated this PR. The mechanism I added initially is now used to figure out a default value for the new path settings. This way users on other platforms can continue using this feature but now with the added benefit of having a more precise control of its behavior.

Scrumplex committed:
fix: only add native library overrides if files exist

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

orowith2os commented:
Will test and provide some quick feedback for Linux shortly.

Scrumplex committed:
fix: add missing header

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

orowith2os commented on CMakeLists.txt:
```diff
@@ -208,6 +208,18 @@ set(Launcher_SUBREDDIT_URL "https://prismlauncher.org/reddit" CACHE STRING "URL
+if(UNIX AND APPLE)
+    set(Launcher_GLFW_LIBRARY_NAME "libglfw.dylib" CACHE STRING "Name of native glfw library")
+    set(Launcher_OPENAL_LIBRARY_NAME "libopenal.dylib" CACHE STRING "Name of native openal library")
+elseif(UNIX)
+    set(Launcher_GLFW_LIBRARY_NAME "libglfw.so" CACHE STRING "Name of native glfw library")
+    set(Launcher_OPENAL_LIBRARY_NAME "libopenal.so" CACHE STRING "Name of native openal library")
```
What happens if the .so files are versioned? Do most (or all) platforms that version them just provide symlinks such that the unversioned path works, and so this will?

Scrumplex replied:
I personally never saw a distro that doesn't have symlinks. This should cover 99% of use-cases. We can expand this later if needed

orowith2os commented on launcher/ui/pages/global/MinecraftPage.ui:
```diff
@@ -214,21 +214,55 @@
+           <widget class="QLabel" name="labelGLFWPath">
+            <property name="text">
+             <string>&amp;GLFW library path</string>
+            </property>
+            <property name="buddy">
+             <cstring>lineEditGLFWPath</cstring>
+            </property>
+           </widget>
+          </item>
+          <item row="2" column="0">
            <widget class="QCheckBox" name="useNativeOpenALCheck">
             <property name="text">
              <string>Use system installation of &amp;OpenAL</string>
             </property>
            </widget>
           </item>
+          <item row="3" column="0">
+           <widget class="QLabel" name="labelOpenALPath">
+            <property name="text">
+             <string>&amp;OpenAL library path</string>
+            </property>
+            <property name="buddy">
+             <cstring>lineEditOpenALPath</cstring>
+            </property>
+           </widget>
+          </item>
+          <item row="1" column="1">
+           <widget class="QLineEdit" name="lineEditGLFWPath">
+            <property name="enabled">
+             <bool>false</bool>
+            </property>
+           </widget>
+          </item>
+          <item row="3" column="1">
+           <widget class="QLineEdit" name="lineEditOpenALPath">
+            <property name="enabled">
+             <bool>false</bool>
+            </property>
+           </widget>
+          </item>
```
Is it possible to have these enter a "Can't find the {name} library automatically! Does it exist?" when it can't be found by Prism? 

Having the text automatically filled out so that the user doesn't have to type it out would be nice too, right now it's just *shown* and can't be edited as-is. Though then auto updating it might be a bit of a pain? Can easily disregard this concern.

Scrumplex replied:
We use placeholder texts in other places as well for pre-set/auto-detected value (check the APIs page)

It's probably gonna auto-detect the right libraries 99% of the time, so we are fine here.

orowith2os reviewed (approved):
Any concerns I might have have been addressed, it all works fine here on the Flatpak, and it feels nice being able to play Minecraft natively on Wayland with ease now. Thanks!

getchoo reviewed (approved):
lgtm

TheKodeToad commented on launcher/MangoHud.cpp:
```diff
@@ -106,4 +116,37 @@ QString getLibraryString()
     return QString();
 }
+
+QString findLibrary(QString libName)
+{
+#ifdef __GLIBC__
+    const char* library = libName.toLocal8Bit().constData();
+
+    void* handle = dlopen(library, RTLD_NOW);
+    if (!handle) {
+        qCritical() << "dlopen() failed:" << dlerror();
+        return {};
+    }
+
+    char path[PATH_MAX];
+    if (dlinfo(handle, RTLD_DI_ORIGIN, path) == -1) {
+        qCritical() << "dlinfo() failed:" << dlerror();
+        dlclose(handle);
+        return {};
+    }
+
+    auto fullPath = FS::PathCombine(QString(path), libName);
+
+    dlclose(handle);
+    return fullPath;
+#else
+    qWarning() << "MangoHud::findLibrary is not implemented on this platform";
+    return {};
+#endif
+}
```
:mango:
I'm pretty sure this should be in a different file.

orowith2os replied:
It was noted by @Scrumplex that it needed to be changed at some point, and that it would be handled later IIRC

Scrumplex replied:
yep. I am already using this file in #1160. I think we can just rename it to something like NativeUtils eventually

TheKodeToad replied:
`NativeHelper` or `NativeUtil(s)` would be good, yes

Scrumplex replied:
I would propose we do this in #1160 as it adds lots of other functions to this file

Scrumplex commented:
reminder not to merge before staging is done

Scrumplex commented on CMakeLists.txt:
creating a conversation to prevent merging

Scrumplex commented:
rebased to develop

Scrumplex commented:
Could someone test this feature on Windows? Obviously, auto-detection of libraries is not implemented there.
